### PR TITLE
Remove the remaining users of config_compilers.xml

### LINKED ISSUE #4136: PFUNIT-based unit tests should use new CMake Macros files, not old config_compilers.xml
In updating my Mac's CIME port to the new CMake Macro-based system, I noticed that `scripts/fortran_unit_testing/run_tests.py` still needs things defined in config_compilers.xml: defining PFUNIT_PATH (and maybe other things?) in the new macros format does not seem sufficient.

For example, if I remove the path to PFUNIT from the cheyenne section in CESM's config_compilers.xml, I get

```
ERROR: PFUNIT_PATH not found for this machine and compiler, with MPILIB=mpi-serial and compile_threaded=FALSE.
You must specify PFUNIT_PATH in config_compilers.xml, with attributes MPILIB and compile_threaded.
```

(I tried this out of the latest CTSM master, with its externals, running `../cime/scripts/fortran_unit_testing/run_tests.py --build-dir unit_tests.temp` from CTSM's `src` directory.)

Since I think we want to eventually remove the need for config_compilers.xml (I assume?), I'm thinking this run_tests.py script should be updated to use the new macros-based method.

By the way, this exercise made me explore the new CMake Macro-based system in more detail. I feel like this is a definite improvement over the previous config_compilers.xml. Thank you @jgfouca and @jedwards4b for all of your work on this!

### PR DESCRIPTION
**Pull request by jgfouca**
The Compilers XML object should not even be constructed unless
the env var enabling the deprecated macro system is on.

Test suite: scripts_regression_tests
Test baseline:
Test namelist changes:
Test status: bfb

Fixes #4136 

User interface changes?:

Update gh-pages html (Y/N)?:

### CODE REVIEW

**Comment by jgfouca** [comment]
This is a bit of a work in progress. Without a proper case to work off of, interrogating CIME for build info is a bit awkward.

File: scripts/lib/CIME/BuildTools/configure.py [hunk 1]
```diff
@@ -79,6 +79,31 @@ def configure(
                     new_cmake_macros_dir, os.path.join(output_dir, "cmake_macros")
                 )
 
+            if form == "Makefile":
+                # Use the cmake macros to generate the make macros
```
Comment (jgfouca) [inline scripts/lib/CIME/BuildTools/configure.py:83]: This feel tacky. Fortunately, it appears as though using `configure` to create Makefile macros is very uncommon. With this PR, the only remaining use is in `tools/mapping/gen_domain_files/test_gen_domain.sh`

File: scripts/lib/CIME/tests/utils.py [hunk 2]
```diff
@@ -353,34 +353,6 @@ def assert_variable_matches(self, var_name, regex, env=None, var=None):
         self.parent.assertRegexpMatches(self.query_var(var_name, env, var), regex)
 
 
-def get_macros(macro_maker, build_xml, build_system):
```
Comment (jgfouca) [inline scripts/lib/CIME/tests/utils.py:356]: This was a duplicate definition.

File: tools/cprnc/README [hunk 3]
```diff
@@ -12,30 +12,19 @@ your case directory.
 Quick Start Guide:
 ------------------
 
-On cime supported systems you can generate a Macros file using the following
+On cime supported systems you can generate a cmake Macros file using the following
 (assuming you are running the command from the directory cime/tools/cprnc):
 
-CIMEROOT=../.. ../configure --macros-format=Makefile --mpilib=mpi-serial
```
Comment (jgfouca) [inline tools/cprnc/README:18]: We want users to use CMake. Unfortunately, there is still a Makefile build of cprnc in `tools/mapping/gen_domain_files/test_gen_domain.sh`, so I couldn't get rid of the Makefile.

File: scripts/lib/CIME/tests/test_sys_unittest.py [hunk 4]
```diff
@@ -7,30 +7,32 @@
...
-            "PFUNIT_PATH", attributes=attrs
-        )
-        if pfunit_path is None:
-            return False
+            compiler = self.TEST_COMPILER
+
+        mach = self.MACHINE.get_machine_name()
+        cmake_macros_dir = Files().get_value("CMAKE_MACROS_DIR")
+
+        machine_macro = os.path.join(cmake_macros_dir, "{}_{}.cmake".format(compiler, mach))
```
Comment (jgfouca) [inline scripts/lib/CIME/tests/test_sys_unittest.py:29]: Kinda hacky.

File: scripts/lib/CIME/tests/test_sys_cmake_macros.py [hunk 5]
```diff
@@ -28,6 +29,9 @@ def xml_to_tester(self, xml_string):
     def setUp(self):
         super().setUp()
 
+        if "CIME_NO_CMAKE_MACRO" not in os.environ:
```
Comment (jgfouca) [inline scripts/lib/CIME/tests/test_sys_cmake_macros.py:32]: Don't test the deprecated system unless the user indicates they want the deprecated system in their env.

File: scripts/lib/CIME/tests/test_sys_full_system.py [hunk 6]
```diff
@@ -12,8 +12,7 @@
 class TestFullSystem(base.BaseTestCase):
     def test_full_system(self):
         # Put this inside any test that's slow
-        if self.FAST_ONLY:
-            self.skipTest("Skipping slow test")
+        self.skipTest("Skipping slow test")
```
Comment (jgfouca) [inline scripts/lib/CIME/tests/test_sys_full_system.py:15]: Need to revert this change.

File: scripts/lib/CIME/BuildTools/configure.py (hunk 1 above)
Comment (jedwards4b) [reply scripts/lib/CIME/BuildTools/configure.py:83]: I don't see any reason that we cannot replace this (in  test_gen_domain.sh) with a cmake build - do you?  That would allow us to remove this ```if form```?

**Review by jedwards4b — APPROVED** [review]

File: scripts/lib/CIME/BuildTools/configure.py (hunk 1 above)
Comment (jgfouca) [reply scripts/lib/CIME/BuildTools/configure.py:92]: @jedwards4b , it looks like the source code in `tools/mapping/gen_domain_files/src/*` requires a Makefile build system, so we'd need to change that in order to be able to remove Makefile mode from `configure`.

**Comment by jgfouca** [comment]
As a side note, it looks like there was a bit of confusion in the code as to where local cmake macros should be located.

The test `CIME.tests.test_sys_create_newcase.TestCreateNewcase.test_ka_createnewcase_extra_machines_dir` was assuming they would live directly under extra_mach_dir, e.g. `$extra_mach_dir/*.cmake`. `case_setup` was looking for local macros in `$extra_mach_dir/cmake_macros/*.cmake`. This was an easy fix, but I'll note that `case_setup` also looks for local macros in `$HOME/.cime/*cmake`. So, we expect extra-mach-dir to provide cmake macros under the directory `cmake_macros`, but we do not expect this in `$HOME/.cime` cmake macros which are expected to live directly in the directory, so things are not consistent between these two approaches. Was this a deliberate choice or accident? @jedwards4b @billsacks

File: scripts/fortran_unit_testing/run_tests.py [hunk 7]
```diff
@@ -425,19 +421,24 @@ def _main():
     )
     machspecific = EnvMachSpecific(build_dir, unit_testing=True)
 
-    fake_case = FakeCase(compiler, mpilib, debug, comp_interface)
+    fake_case = FakeCase(compiler, mpilib, debug, comp_interface, threading=use_openmp)
     machspecific.load_env(fake_case)
+    # JGF bug? original contents of cmake_args is lost
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/fortran_unit_testing/run_tests.py:426]: In the previous version cmake_args was not an input argument so I don't think this condition was ever satisfied, but now that it is, this does appear to be a bug.

[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/fortran_unit_testing/run_tests.py]: @jedwards4b , I pushed what I think is the appropriate fix for this.

**Comment by billsacks** [comment]
> As a side note, it looks like there was a bit of confusion in the code as to where local cmake macros should be located.
> 
> The test `CIME.tests.test_sys_create_newcase.TestCreateNewcase.test_ka_createnewcase_extra_machines_dir` was assuming they would live directly under extra_mach_dir, e.g. `$extra_mach_dir/*.cmake`. `case_setup` was looking for local macros in `$extra_mach_dir/cmake_macros/*.cmake`. This was an easy fix, but I'll note that `case_setup` also looks for local macros in `$HOME/.cime/*.cmake`. So, we expect extra-mach-dir to provide cmake macros under the directory `cmake_macros`, but we do not expect this in `$HOME/.cime` cmake macros which are expected to live directly in the directory, so things are not consistent between these two approaches. Was this a deliberate choice or accident? @jedwards4b @billsacks 

I ran into this confusion, too. Once I figured out what was going on, I came to feel that the way it's done for the extra_machines_dir (having a cmake_macros subdirectory) makes sense in that it parallels what's done for the actual machines directory, though I didn't (and still don't) feel strongly one way or the other. (This led me to add clarification in the documentation in #4137 ). I can see arguments either way for `.cime`. I do see that it could be more clear if these two were consistent. In that case, I don't have much preference as to whether they are both done the way `.cime` is or the way `extra_machines_dir` is done, though if I had to choose, I'd vote for using a `cmake_macros` subdirectory in both cases.

**Comment by jedwards4b** [comment]
I think that it makes most sense to have .cime/cmake_macros/*.cmake

**Comment by jgfouca** [comment]
@jedwards4b , @billsacks , I agree that it makes sense to have `~/.cime/cmake_macros`. Unfortunately, this change would break backwards compatibility and I know some E3SM devs that are always using custom macros, so I am a bit hesitant to just make the change here. Perhaps I will take this on in a separate PR.

**Comment by billsacks** [comment]
> I agree that it makes sense to have `~/.cime/cmake_macros`. Unfortunately, this change would break backwards compatibility and I know some E3SM devs that are always using custom macros, so I am a bit hesitant to just make the change here. Perhaps I will take this on in a separate PR.

That makes sense - thanks for all of this.
